### ISSUE TITLE
Two errors on Appsody project creation

### ISSUE THREAD

**Issue by makandre** [eclipse/codewind#1073]
**Codewind version:** 0.6.0
**OS:**

**Che version:**
**IDE extension version:**
**IDE version:**
**Kubernetes cluster:**

**Description:**

The symptom observed here is same as https://github.com/eclipse/codewind/issues/936 but this is on Che and scenario is different.  The following 2 errors are seen when re-opening a closed Che workspace:

![image](https://user-images.githubusercontent.com/17596240/68511779-ae170f80-0244-11ea-8f77-8d3380eb5f58.png)

**Steps to reproduce:**

1. create an appsody project
2. stop the Che workspace
3. come back to the workspace, when Theia comes up, expand the Codewind explorer
   - at this point you should see Codewind (PFE) is starting
4. once PFE starts, the existing project is re-bound to Codewind
   - this is when the 2 errors show up

**Workaround:**

**Comment by makandre** [comment]
Not sure why we don't see this for Codewind projects. One difference I can see is when projects are re-bound to Codewind, Appsody projects pretty quickly goes to `running` state, since they don't need to go through "build" stage, whereas Codewind projects goes to `building`

I tried with appsody and codewind nodejs projects

**Comment by makandre** [comment]
Briefly chatted with @tetchel, he suggested maybe it just takes more time for Turbine to become aware of appsody projects.

That make sense, there's more stuff to run through when Turbine first initializes a project handler for appsody, and the `fwdata` only gets written after that.

I wonder, if we had Turbine emit a new event after it writes the `fwdata`, could we delay the get logs and capabilities to that point when we know for sure those calls will succeed?

**Comment by rajivnathan** [comment]
Both the capabilities and logs API are expected to return errors if the project has not started yet. I believe we've seen this problem in the past where portal tries to call those APIs prematurely? @tobespc Does that ring a bell?

**Comment by makandre** [comment]
Tho I think in this case it's VSCode/Theia that's making those calls prematurely

**Comment by tetchel** [comment]
Yes andrew is correct

**Comment by tetchel** [comment]
#1100 now happening with regular codewind projects too so we may want to increase priority if @elsony agrees

**Comment by jagraj** [comment]
Looks like this seems to be regression, can we make this issue as "stopship" or hot for 0.6.0 release.?

**Comment by elsony** [comment]
This one looks like to be the same error as #936 that portal owned before. The problem was due to build was not done when the logs are being request. This one looks like to be a different variant or that problem that shows up on a restart of PFE instead of first time creation. Therefore, I am transferring this to portal.

**Comment by makandre** [comment]
Seeing this more and more now with hybrid testing

**Comment by keithchong** [comment]
I see this locally too, using Codewind 0.6.1.

**Comment by tetchel** [comment]
I see this on every new project creation in hybrid

**Comment by tobespc** [comment]
this is a dupe of #1471 , moving to verify to test once 1471 has landed

**Comment by tetchel** [comment]
I don't see how this is a dupe of 1471

**Comment by malincoln** [comment]
@makandre and @jagraj can verify once PR for 1471 is merged.

**Comment by malincoln** [comment]
@makandre @jagraj 1471 is now closed.  Please verify.  Thanks

**Comment by makandre** [comment]
I do not see the 2 errors on workspace re-open.  However, I see them pretty much on every project creation now.

pls see https://github.com/eclipse/codewind/issues/1073#issuecomment-552048475 for some background

**Comment by malincoln** [comment]
@makandre do you have any logs that you can provide to @tobespc to investigate further?

**Comment by jopit** [comment]
I can reproduce this on both local and hybrid.  Here's the PFE log from local.

[pfe.txt](https://github.com/eclipse/codewind/files/3974341/pfe.txt)

**Comment by rajivnathan** [comment]
Briefly chatted with Andrew. Think his earlier comment is how we can fix this. It'd require changes on Turbine and the IDEs cc @elsony @jopit 

Adding iterative-dev and IDE area labels and downgrading from stopship to hot because it does not lead to functional problems as Andrew pointed out.

> Briefly chatted with @tetchel, he suggested maybe it just takes more time for Turbine to become aware of appsody projects.
> 
> That make sense, there's more stuff to run through when Turbine first initializes a project handler for appsody, and the `fwdata` only gets written after that (https://github.com/eclipse/codewind/blob/master/src/pfe/file-watcher/server/src/controllers/projectsController.ts#L341).
> 
> I wonder, if we had Turbine emit a new event after it writes the `fwdata`, could we delay the get logs and capabilities to that point when we know for sure those calls will succeed?

**Comment by tetchel** [comment]
I believe the one functional problem we should see from this is that unsupported capabilities are enabled (ie, restart, app monitor). If there's an error getting the capabilities (as seen here) then the extension falls back to assuming all capabilities are available.

**Comment by tobespc** [comment]
I like @rajivnathan suggestion but rather than wait for a new event to be triggered, can we not use the existing one to say a build has started ?  I guess if autobuild is off though then we wouldn't get a build event. So I've argued against myself :-) .  

so +1 to @rajivnathan for that idea

**Comment by jopit** [comment]
I've updated the title to more accurately reflect the problem.

**Comment by elsony** [comment]
@tetchel What event is the IDE listening for before calling the capabilities?  We currently have a project creation successful event (not just the acknowledgement of project creation) sent out when the project creation is completed.  Before that event is issued, no operation can be done by the project on the Turbine layer.  If we happens to be sending the project creation success event too early before the Appsody initialization is completed, then we should investigate on Turbine/Appsody extension to see why that is the case and fix that.  If the IDE happens to be issuing events before the the project creation success event occurs, then that should be fixed on the IDE side to prevent that. Either way, the writing of the fwdata should be irrelevant and considered as internal operation of Turbine so I don't think adding a new notification on when we write the fwdata is the correct solution.

**Comment by tetchel** [comment]
i am using `projectBind` because `projectCreation` takes way too long to be emitted (https://github.com/eclipse/codewind/issues/747)

**Comment by makandre** [comment]
If `projectCreation` takes too long, then what about the `newProjectAdded` event?

**Comment by tetchel** [comment]
Never heard of that one 🙃

**Comment by jopit** [comment]
Isn't that event emitted on the filewatcher socket and not on the ui socket?

**Comment by tetchel** [comment]
for the most part, Portal propagates FW events to its clients (the IDEs) but it's likely that one isn't, since each event has its own 'forwarding function'.

**Comment by makandre** [comment]
I think it is propagated:

https://github.com/eclipse/codewind/blob/0.7.0/src/pfe/portal/modules/FileWatcher.js#L473

**Comment by jopit** [comment]
Yeah, it is propagated.  Here's the sequence of events that are emitted when creating a simple appsody node project
```
2019-12-17T17:57:41.568Z: projectStatusChanged
2019-12-17T17:57:41.657Z: projectBind
2019-12-17T17:57:41.684Z: projectStatusChanged
2019-12-17T17:57:41.701Z: newProjectAdded
2019-12-17T17:57:41.816Z: projectWatchStatusChanged
2019-12-17T17:57:41.990Z: projectStatusChanged
2019-12-17T17:57:41.998Z: projectValidated
2019-12-17T17:57:42.992Z: projectStatusChanged
2019-12-17T17:57:43.029Z: projectLogsListChanged
2019-12-17T17:57:46.448Z: projectStatusChanged
2019-12-17T17:58:13.110Z: projectCreation
2019-12-17T17:58:14.425Z: projectStatusChanged
```

**Comment by yangcao77** [comment]
The `newProjectAdded` is not a proper event for ide to listen to. It's for filewatcherd, and it's not emitted for queued projects.

**Comment by elsony** [comment]
ProjectBind is definitely not good enough. ProjectBind get issued as long as portal issue a bind command to Turbine. There is no guarantee what code has turbine goes through yet so I think we kind of luck out to get things working today if we are relying on ProjectBind.  ProjectCreation is the original notification that is defined for IDE to safely start doing operations on the project.  So that is the safest one to use so far since that is the one that we know all project operations/APIs will be ready by that time. If we want something earlier than that, then we'll probably need a new notification, e.g. something like a "project ready", for that purpose.  However, we'll need to understand what are the possible operations that the IDE can be calling on a project to decide when and where is a safe place for us to make the cut to issue the new notification. We may ended up using one of the existing notification if it works out, but we need to understand what potential operations that IDE will call when you receive that "project ready" notification.

**Comment by tetchel** [comment]
fixing #747 would likely make the projectCreation event suitable to use for this as is intended

**Comment by elsony** [comment]
One thing that you mentioned earlier is ProjectCreation takes a long time. A big chunk of the time is on building the build and runtime containers.  Therefore, even if we change the ProjectCreation to the end of building the build and runtime container, it still take long to issue that event.  Also, even if build and runtime containers are build, some operations like getting the URL will not be ready yet.  Therefore, to fix it properly, we'll still need input on how the IDE interrupt that "project ready" event and what operations will you potentially call after you receive the "project ready" notification so we can decide when and where we can safely issue that. We can track the work in this issue or #747. I think both item may end up to be the same fix.

**Comment by tobespc** [comment]
We have discussed this in the past but bringing it up again here.   Why can we not implement the capabilities function in portal ?   Is there anything that requires build specific (that only Turbine knows about) data to be used ?  Are the capabilities bound to language and type ?   If we don't need to wait for Turbine to give us this data then we should move this to portal.

**Comment by makandre** [comment]
But the other API called is getting the logs, think that one is squarely in Turbine side.

**Comment by rajivnathan** [comment]
Regardless of whether it's the capabilities or logs API we're talking about the issue is that the calls are happening too early. If the same call is made later (when the project metadata is initialized) then everything works just fine. Think the solution is to use the ProjectCreation event or create a new event to signal when Turbine APIs can be invoked.

@tobespc to answer your question, cannot move implementation of those APIs to portal because some are inherently at the build container level and require Turbine to provide file paths, etc. eg. logs

**Comment by jopit** [comment]
From the user's point of view, we want the new project to show up in the IDE as soon as possible, which means having it appear before it's properly ready.  Currently vscode is keying off the 'projectBind' event, and eclipse is keying off the initial 'projectStatusChanged' event.

It would be a better design if there was some other event that was emitted when PFE first starts the process of creating a new project, instead of a 'projectStatusChanged' event with an unknown (to the IDEs) project id.

Eclipse is also seeing these two errors, but is not surfacing them to the user.  Changing eclipse to key off the 'newProjectAdded' does prevent the errors.

**Comment by rajivnathan** [comment]
These errors don't prevent the IDE from showing the projects earlier though? It can still appear before it's properly ready but you just can't perform actions like run/debug (capabilities) or show logs before you have containers for the projects.

**Comment by jopit** [comment]
The errors don't prevent the IDEs from showing the projects early.

And yes, actions like run/debug are different from actions like showing the logs.  A developer might want to stream the logs from the initial build even though the project isn't ready to run -- if nothing else it's a way to see that things are actually happening, especially for projects that take a long time before they're ready.  An event to announce those APIs are available probably makes sense.

**Comment by elsony** [comment]
yes, so the only missing piece of info is to know which are the list of API/actions that IDE expect to work with that notification so we can pick the point to send that new notification.

**Comment by elsony** [comment]
Based on the discussions, here is the proposed fix to this problem:
1. Turbine will create a new notification, e.g. projectCapabilitiesReady (final name TBD), after Turbine has handled the project creation up to a stage that the capability info is ready, i.e. when Turbine has finished creation of metadata so Turbine has registered that project.
2. IDE can listen to the projectCapabilitiesReady event to decide whether it can call the capability API to find out the capabilities and start building the menu.
3. For the logs, the IDE will wait for the existing projectLogsListChanged event before calling the get logs operation.
4. IDE will wait for the projectCreation success event before calling all other operations on the project.

**Comment by elsony** [comment]
@yangcao77 Can you work on the Turbine work, i.e. item #1 on my previous comment.

**Comment by elsony** [comment]
Turbine work for item #1 has been completed.

**Comment by malincoln** [comment]
@elsony do we already have someone assigned from IDE to complete step 2?  Are they using same issue or separate one?

**Comment by elsony** [comment]
I had a meeting with them last Fri. to talk about the design. I assume @eharris369  will be working on the Eclipse side, @tetchel will be working on the VSCode side and using this item to track the work on the IDE side as well. @jopit Can you confirm?

**Comment by elsony** [comment]
It turns out we also need https://github.com/eclipse/codewind/pull/1728 which is currently waiting for review from portal.

**Comment by tetchel** [comment]
Blocked by https://github.com/eclipse/codewind/issues/1752

**Comment by malincoln** [comment]
1752 has been merged into 0.8.0 branch.  Master branch fix in progress

**Comment by eharris369** [comment]
Eclipse PRs:
- 0.8.0: https://github.com/eclipse/codewind-eclipse/pull/529
- master: https://github.com/eclipse/codewind-eclipse/pull/530

**Comment by malincoln** [comment]
@makandre this is now ready to verify.  Thanks

**Comment by makandre** [comment]
Verified I don't see the errors anymore
